fix the usage of is_alias_of in ddp reducer

zhaojuanmao opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#52328 fix the usage of is_alias_of in ddp reducer**

is_alias_of return true if one tensor is sub_tensor of another tensor. It is not properly used in DDP reducer right now. Introduce a new helper function in c10d to check whether two tensors share the same storage.

Differential Revision: [D26470577](https://our.internmc.facebook.com/intern/diff/D26470577/)

zhaojuanmao committed:
fix the usage of is_alias_of in ddp reducer

is_alias_of return true if one tensor is sub_tensor of another tensor. It is not properly used in DDP reducer right now. Introduce a new helper function in c10d to check whether two tensors share the same storage.

Differential Revision: [D26470577](https://our.internmc.facebook.com/intern/diff/D26470577/)

[ghstack-poisoned]

rohan-varma commented on torch/lib/c10d/reducer.cpp:
```diff
@@ -405,7 +405,7 @@ void Reducer::mark_variable_ready_dense(VariableIndex index) {
       // to bucket_view. If grad has already been set as views of buckets in
       // previous iterations, no copy is needed.
-      if (!grad.is_alias_of(bucket_view)) {
+      if (!c10d::shareSameStorage(grad, bucket_view)) {
```
Can we add a unittest that would've failed if we used `is_alias_of` (presumably involving using ddp with gradient_as_bucket_view with same-storage tensors)?

zhaojuanmao replied:
@rohan-varma it is not failing right now, but I think it is not properly used, see my replies to @albanD's comments.

albanD commented on torch/lib/c10d/Utils.hpp:
```diff
@@ -392,6 +392,15 @@ std::vector<T*> getDataPointers(const std::vector<at::Tensor>& tensors) {
 }
 
+// Helper to check whether t1 and t2 share the same storage
+inline bool shareSameStorage(const at::Tensor& t1, const at::Tensor& t2) {
```
Storage has a very specific meaning in pytorch as Storage is a type of elementary object so I don't think you want to use it here.

Do you actually want to check if the two Tensors can potentially look at the same memory? If so, we already have function to check if Tensors alias each other.

Otherwise, could you explain what this function is checking for exactly? In particular, it is very surprising that it checks for things like numel and stride but not size.

zhaojuanmao replied:
@albanD thanks for commenting! 

@mcarilli reminded me that "is_alias_of()" returning true as long as the underlying storage is the same for two tensors (even if they view different regions of the storage). 

In this case, "is_alias_of()" seems not to be the right method used in DDP. In DDP, we allocate a large tensor e.g., called Bucket. we've let grad1 point to one slice of the Bucket called bucket_view_1, and let grad2 point to another slice of the Bucket called bucket_view_2.  

What we want is: shareSameStorage(grad2, bucket_view_1) return False, However, grad2.is_alias_of(bucket_view_1) will return True. 

That is why I'm trying to write this method to make sure shareSameStorage(grad2, bucket_view_1) return False. 

Does it make sense to you?  agree that the naming is not proper, also I'm not confident that what I'm checking is the right way. That is why I'm requesting your review and feedback about how to write the right helper method to check two tensors are the same (or, am I overthinking?  Will t1==t2 work..?)

albanD replied:
Could you give a full description on where this function should return True and False?

zhaojuanmao committed:
Update on "fix the usage of is_alias_of in ddp reducer"


is_alias_of return true if one tensor is sub_tensor of another tensor. It is not properly used in DDP reducer right now. Introduce a new helper function in c10d to check whether two tensors share the same storage.

Differential Revision: [D26470577](https://our.internmc.facebook.com/intern/diff/D26470577/)

[ghstack-poisoned]

zhaojuanmao commented on torch/lib/c10d/reducer.cpp:
```diff
@@ -422,7 +422,7 @@ void Reducer::mark_variable_ready_dense(VariableIndex index) {
       // to bucket_view. If grad has already been set as views of buckets in
       // previous iterations, no copy is needed.
-      if (!grad.is_alias_of(bucket_view)) {
+      if (!c10d::viewSameMemory(grad, bucket_view)) {
```
@albanD if I add a log before this line 
```
      if (process_group_->getRank() == 0 && num_iterations_ == 2) {
        for (size_t i = 0; i < replica.bucket_views_in.size(); i++) {
          std::cout << bucket_index.intra_bucket_index << "th grad is alias of "
            << i << "th bucket view, " << grad.is_alias_of(replica.bucket_views_in.at(i))
            << ", they view the exact same memory: "
            << c10d::viewSameMemory(grad, replica.bucket_views_in.at(i))
            << "\n";
        }
      }
```
and run a unit test with grad_as_bucket_view = true,  it will print out logs like this 
```
0th grad is alias of 0th bucket view, 1, they view the exact same memory: 1
0th grad is alias of 1th bucket view, 1, they view the exact same memory: 0
0th grad is alias of 2th bucket view, 1, they view the exact same memory: 0
1th grad is alias of 0th bucket view, 1, they view the exact same memory: 0
1th grad is alias of 1th bucket view, 1, they view the exact same memory: 1
1th grad is alias of 2th bucket view, 1, they view the exact same memory: 0
2th grad is alias of 0th bucket view, 1, they view the exact same memory: 0
2th grad is alias of 1th bucket view, 1, they view the exact same memory: 0
2th grad is alias of 2th bucket view, 1, they view the exact same memory: 1
```
I’m expecting 0th grad will be alias of 0th bucket_view only. But the log shows the 0th grad is alias of all bucket_views in the bucket buffer tensor. 

Although it is not resulting in numeric error for current case we've known. But we feel like is_alias_of is not what we want originally. So trying to write a helper to check whether the two tensors are actually viewing the exact same memory.

albanD replied:
And what should it return for partial overlap? True or False?

zhaojuanmao replied:
we want it to return False if two tensors are partially overlapped. Basically we want two tensors to return True only when they are exactly the same. (Will tensor1==tensor2 work...?)

albanD replied:
In that case, yes you can check for equality. As Tensor in c++ are shared pointers and a single object in python.

Note though that `t1 == t1.view_as(t1)` will be false because the result of the view op is a new Tensor. That is what you want right?
